pkg/kubelet/cm/pod_container_manager_linux.go

```diff
@@ -108,10 +114,101 @@ func (m *podContainerManagerImpl) GetPodContainerName(pod *api.Pod) string {
 	return path.Join(parentContainer, podContainer)
 }
 
+// Scans through all the subsystems pod cgroups directory.
+// Will also scan through the container cgroups that still exist under
+// the pod cgroup and haven't been Garbage Collected yet
+func (m *podContainerManagerImpl) getPIDsToKill(podCgroup string) []int {
+	// Get a list of processes that we need to kill
+	pidsToKill := sets.NewInt()
+	var pids []int
+	for _, val := range m.subsystems.MountPoints {
+		dir := path.Join(val, podCgroup)
+		_, err := os.Stat(dir)
+		if os.IsNotExist(err) {
+			// The subsystem pod cgroup is already deleted
+			// do nothing, continue
+			continue
+		}
+		// Get a list of pids that are still charged to the pod's cgroup
+		pids, err = getCgroupProcs(dir)
+		if err != nil {
+			continue
+		}
+		pidsToKill.Insert(pids...)
+
+		// WalkFunc which is called for each file and directory in the pod cgroup dir
+		visitor := func(path string, info os.FileInfo, err error) error {
+			if !info.IsDir() {
+				return nil
+			}
+			pids, err = getCgroupProcs(path)
+			if err != nil {
+				return err
+			}
+			pidsToKill.Insert(pids...)
+			return nil
+		}
+		// Walk through the pod cgroup directory to check if
+		// container cgroups haven't been GCed yet. Get attached processes to
+		// all such unwanted containers under the pod cgroup
+		err = filepath.Walk(dir, visitor)
+	}
+	return pidsToKill.List()
+}
+
+// Scan through the whole cgroup directory and kill all processes either
+// attached to the pod cgroup or to a container cgroup under the pod cgroup
+func (m *podContainerManagerImpl) tryKillingCgroupProcesses(podCgroup string) error {
+	pidsToKill := m.getPIDsToKill(podCgroup)
+	// No pids charged to the terminated pod cgroup return
+	if len(pidsToKill) == 0 {
+		return nil
+	}
+
+	var errlist []error
+	// os.Kill often errors out,
+	// We try killing all the pids multiple times
+	for i := 0; i < 5; i++ {
+		if i != 0 {
+			glog.V(3).Infof("Attempt %v failed to kill all unwanted process. Retyring", i)
```

Comment:
it would also be good if the log messages said which pod it was cleaning up.  so "pod container manager: pod <uid> failed to kill all unwanted processes".